### ISSUE TITLE
speculative_generate_step diverges from plain generation at temp=0 with num_draft_tokens=4 (dense model, no hybrid layers)

### ISSUE THREAD

**Issue by h9q2cyxvgm-ui** [ml-explore/mlx-lm#1470]
### Summary

`speculative_generate_step` (and `generate()`/`stream_generate` with `draft_model=`) produces output that diverges from plain greedy generation at `temp=0` with `num_draft_tokens=4`, on a plain dense model + a small peer draft model — no hybrid/recurrent architecture involved, no custom code. Speculative decoding at temp=0 is supposed to be lossless relative to plain generation (that's the whole point — it should only ever produce a different token when the draft model happens to sample the exact same one the target would have), so this looks like a real bug rather than expected variance.

Confirmed on both current `main` and the tip of an unrelated feature branch (`h9q2cyxvgm-ui/mlx-lm@feature/qwen3-5-mtp-speculative-decoding`, see #1468/#1469) — same result either way, so it's not something introduced by that branch's changes.

### Reproduction

```python
import mlx.core as mx
from mlx_lm.generate import generate_step, speculative_generate_step
from mlx_lm.utils import load
from mlx_lm.sample_utils import make_sampler

TARGET = "mlx-community/Qwen3-32B-6bit"
DRAFT = "mlx-community/Qwen3-0.6B-MLX-6bit"   # any small same-family draft
MAX_TOKENS = 150

model, tokenizer = load(TARGET)
draft_model, _ = load(DRAFT)

messages = [{"role": "user", "content": "Explain what a hash table is in two sentences."}]
prompt = mx.array(tokenizer.apply_chat_template(messages, add_generation_prompt=True))
sampler = make_sampler(temp=0.0)

plain = [tok for tok, _ in generate_step(prompt, model, max_tokens=MAX_TOKENS, sampler=sampler)]

spec = []
for tok, _, from_draft in speculative_generate_step(
    prompt, model, draft_model, max_tokens=MAX_TOKENS, num_draft_tokens=4, sampler=sampler
):
    spec.append(tok)

print(plain == spec)  # False
for i, (a, b) in enumerate(zip(plain, spec)):
    if a != b:
        print(i, a, b)
        break
```

### Result

First divergence at token index 91: plain generates token `429` (`" that"`), speculative generates `1246` (`" how"`) — a genuine content divergence partway through generation, not just a length/edge-case difference.

### Notes

- `num_draft_tokens=1`, `2`, `3` all matched plain generation exactly in my testing on the same prompt/models — only `4` reproduced the mismatch so far. Haven't swept further values or other prompts to find the exact boundary.
- Target and draft are both plain dense Qwen3 models (`model_type: qwen3`), no `ArraysCache`/hybrid recurrent layers, so this is unrelated to #1462 or the ArraysCache cache-rewind work in #1468.
- Happy to help narrow this down further (e.g. bisect where exactly in the accept/reject/rewind logic the divergence originates) if useful — ran into it incidentally while benchmarking a different change and haven't dug into `speculative_generate_step`'s internals to find the root cause yet.

### Environment

- `mlx-lm` (local build from `main` @ time of filing, mlx `0.31.2`)
- macOS 26.4.1, Apple Silicon (M5 Max, 128GB)
- Python 3.13

**Comment by rajanshxrma** [comment]
Root cause: this looks like a bfloat16-precision tie between competing tokens, not a logic bug — `speculative_generate_step`'s accept/reject/rewind code is provably correct.

Verified line-by-line: `generate.py:622-634` is a pure greedy argmax-equality test, and the cache-rewind math at `589-591` is `num_draft - n` for every `n`, no special-casing. So the divergence isn't in the speculative algorithm itself.

I couldn't run your 32B model (M1/16GB here), so this doesn't reproduce *your* exact instance — but on a small Qwen3-4B/0.6B pair (same architecture family) I found the same *mechanism*, which strongly indicates the same root cause. A num_draft sweep (1-8) matched plain greedy exactly through 7, then diverged at 8 — a different threshold than your num_draft=4, and prompt-dependent (6/10 test prompts diverged at num_draft=8, others matched fine), with no consistent block-boundary pattern. That already points away from a fixed architectural threshold and toward something data-dependent.

Reconstructing the target model's KV cache from scratch (mirroring exactly what plain `generate_step` does — pure sequential, one token at a time) and inspecting the raw logits at the actual divergence point: **tokens 1376 and 1887 are an exact bit-level tie at 38.0 in bfloat16** (`mx.eval(l1376 == l1887)` → `True`). The model runs in bfloat16 (8-bit mantissa), so exact ties between plausible next-token candidates aren't rare. Speculative decoding's target scoring (one batched forward over the whole draft block) and plain generation's target scoring (sequential single-token forwards) are mathematically equivalent but take different reduction/kernel paths through matmul/attention — for a value that's already tied at bfloat16 precision, whichever path rounds fractionally lower wins the tie, and that can differ between the batched and sequential paths.

So: `speculative_generate_step`'s losslessness at temp=0 holds for the accept/reject logic, but not strictly for bfloat16-level tie-breaking between batched and sequential scoring — this looks like an inherent precision limitation rather than something fixable in the algorithm itself. I'd frame this as a documentation clarification (spec-decode is lossless up to floating-point non-associativity, not bit-exact) rather than a code fix. Happy to share the repro/probe scripts if useful for confirming this is the same mechanism on your model.

**Comment by h9q2cyxvgm-ui** [comment]
Thanks for the thorough follow-up, this is exactly the kind of root-causing I was hoping someone would have bandwidth for.  The exact-tie-at-bfloat16 finding is a much sharper explanation than anything I had; I'd only gotten as far as "batched vs. sequential scoring diverges, algorithm looks correct" without pinning down *why*.

This also explains a related thing we ran into on a completely different codepath: while building MTP speculative decoding support for Qwen3.6 and Gemma4 in mlx-lm (separate PRs, #1468/#1469), we found the same general symptom, a rare divergence (roughly 1 per 150 to 190 tokens) between speculative and plain generation on quantized models, at temp=0.  We tracked it down to the same *class* of cause empirically (a self-speculative control test using the target model as its own draft, with no MTP or custom code involved at all, still diverged; plain generation alone was confirmed perfectly deterministic run-to-run, ruling out anything random), but didn't go as far as identifying the actual bfloat16-tie mechanism.  Good to have that confirmed independently on a different model family (Qwen3-4B/0.6B vs. our Qwen3.6/Gemma4 testing); it strengthens the case that this is a general property of batched-vs-sequential floating-point reduction, not something specific to one architecture or code path.

Agreed on the framing: this looks like an inherent floating-point non-associativity limitation, not a fixable logic bug; "lossless up to floating-point associativity, not bit-exact" is the right way to describe it.  Would be good to land that as a doc clarification rather than treating this as an open correctness bug.  Happy to link back here if it comes up again in the MTP work.

**Comment by PhilipJohnBasile** [comment]
@h9q2cyxvgm-ui — we hit this same class of divergence on the 4-bit siblings (Qwen3-32B-4bit + 0.6B-4bit draft, `temp=0`, `num_draft_tokens=4`) and ran a per-position logit audit before attributing it to anything algorithmic.

Our divergence was deterministic at token index 23 (baseline `334`, speculative `729`). Replaying the exact position through the sequential path and extracting raw logits: **both tokens logit `33.75`, softmax `0.3828`, `logit_gap = 0.0` exactly** — ranks 1 and 2, byte-identical across 3 repeats.

The mechanism: at `temp=0`, an exact bf16 tie makes argmax implementation-defined. `speculative_generate_step` verifies in a batched target forward (`num_draft+1` wide) while plain `generate_step` runs sequential single-token forwards — same math, different reduction order, so the tie-break can flip between tied ids while staying quality-equivalent. "Lossless" holds up to floating-point tie-break behavior.

A falsifier for your case: replay your forced prefix through the plain sequential path at the exact divergence index and print both candidate tokens' raw logits/probs/ranks. If the gap is ~0.0 with both tokens top-rank, it's this benign class. If the gap is materially nonzero with a clearly dominant baseline token, then it *is* a real correctness bug, and the audit localizes it to accept/verify. (The harness pattern comes from our engine-certification work.)

Separately, we read the current `speculative_generate_step` looking for an independent defect — accept-comparison indexing, the `num_draft+1` logits slice, both rewind lengths, and sampler ordering all check out. We didn't find one.

**Comment by h9q2cyxvgm-ui** [comment]
Ran the falsifier against our own setup and reproduced the exact same mechanism.  Target `mlx-community/Qwen3-8B-6bit`, peer draft `Qwen3-0.6B-MLX-6bit`, temp=0, greedy, unmodified `speculative_generate_step`.

Swept `num_draft_tokens` 1 through 8 across 5 prompts.  ndt=1 through 7 matched plain generation exactly, ndt=8 diverged at token index 99.  Rather than reconstructing the sequential path after the fact, I captured logprobs live during the original plain generation run, so the divergence-position numbers below are exactly what plain generation itself computed, not an approximation.

At index 99: baseline (plain) token id `3070` (`' **'`), logprob `-1.000000`, rank 0.  Speculative token id `11162` (a stray UTF-8 continuation byte), logprob `-1.000000`, rank 1.  Gap = 0.0 exactly, both candidates tied at the top.

Same mechanism, third independent model family now (Qwen3-4B/0.6B, Qwen3-32B/4bit, and now Qwen3-8B/0.6B), same signature: exact tie, adjacent ranks, zero gap.  I think this closes it out as a genuine, well-corroborated floating-point non-associativity limitation rather than an open correctness question.  Fully on board with the doc-clarification framing over a code fix.

**Comment by rajanshxrma** [comment]
Opened a documentation-only PR to capture this: #1592. Added a `Note:` to the `speculative_generate_step` docstring explaining that at `temp=0` the accept/reject logic is lossless in exact arithmetic, but exact-tied logits (common in bfloat16) can round differently between the batched draft-scoring pass and plain generation's sequential scoring — expected floating-point non-associativity, not a correctness bug. Thanks @h9q2cyxvgm-ui and @PhilipJohnBasile for the cross-model confirmation, that's what made this clear enough to write down.

**Comment by PhilipJohnBasile** [comment]
@rajanshxrma Thanks for turning this into #1592 — that framing matches what the cross-model repro showed. Exact-tied logits in bf16 rounding differently between the batched draft-scoring pass and sequential scoring explains everything I saw, and it correctly leaves the accept/reject logic alone as lossless in exact arithmetic.

No objection to closing this once the docstring note lands; "expected floating-point non-associativity, documented" is the right resolution rather than a code change.

**Comment by rajanshxrma** [comment]
@PhilipJohnBasile Really appreciate you digging in and running your own repro on this — another independent confirmation, on top of what you and h9q2cyxvgm-ui both found, is what nailed down that this is bf16 tie-breaking and not something model-specific. Agreed on the resolution: once the docstring note in #1592 lands, this is "expected floating-point non-associativity, documented" and #1470 can close. Thanks for digging into it with me.
